storage/couchdb/couchdb.go

```diff
@@ -0,0 +1,157 @@
+package couchdb
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/docker/go-connections/tlsconfig"
+	"github.com/flimzy/kivik"
+	couch "github.com/go-kivik/couchdb"
+	"github.com/go-kivik/couchdb/chttp"
+	"github.com/sirupsen/logrus"
+)
+
+// ID of a document in CouchDB; includes Rev for updating an existing
+// document
+type ID struct {
+	ID  string `json:"_id,omitempty"`
+	Rev string `json:"_rev,omitempty"`
+}
+
+// Timing can be embedded into other modesl to add time tracking fields
+type Timing struct {
+	CreatedAt time.Time `json:"created_at"`
+	UpdatedAt time.Time `json:"updated_at"`
+	DeletedAt time.Time `json:"deleted_at"`
+}
+
+func createConnection(tlsOpts tlsconfig.Options, host string, username, password string) (*kivik.Client, error) {
+	// remove username and password from URL to avoid authentication step
+	// before we can set the TLS transport
+	_host, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+	_host.User = nil
+
+	logrus.Debugf("attempting to connect %s to host %s", username, _host.String())
+
+	client, err := kivik.New(context.TODO(), "couch", _host.String())
+	if err != nil {
+		return nil, err
+	}
+
+	t, err := tlsconfig.Client(tlsOpts)
+	if err != nil {
+		return nil, err
+	}
+	// We have to use the original CipherSuites to be able to create a
+	// connection with CouchDB
+	t.CipherSuites = tls.Config{}.CipherSuites
+
+	setXport := couch.SetTransport(&http.Transport{
+		TLSClientConfig: t,
+	})
+
+	if err = client.Authenticate(context.TODO(), setXport); err != nil {
+		return nil, err
+	}
+
+	if username != "" {
+		var basicAuth chttp.Authenticator = &chttp.BasicAuth{
```

Comment:
For the setting of the transport I followed an example shown in the Kivik library:

https://github.com/go-kivik/couchdb/blob/master/auth.go#L41

```
// Example:
//
//     setXport := couchdb.SetTransport(&http.Transport{
//         // .. custom config
//     })
//     client, _ := kivik.New( ... )
//     client.Authenticate(setXport)
```

The client.Authenticate(setXport) doesn't seem to do much except for setting the transport:

First: https://github.com/go-kivik/couchdb/blob/master/auth.go#L17
Then: https://github.com/go-kivik/couchdb/blob/master/auth.go#L33

The 2nd authenticate in my code is this one:

```               if err = client.Authenticate(context.TODO(), basicAuth); err != nil {
                        return nil, err
                }
```

This is the implementation of the Authenticate() function:

https://github.com/go-kivik/couchdb/blob/master/chttp/auth.go#L45

It sets the transport at the end for future use: https://github.com/go-kivik/couchdb/blob/master/chttp/auth.go#L71

I will leave some comments in the code regarding that.
The function then returns a client object that is set up to use for accessing the DB.

I can remove the check for username to require basic authentication all the time, not just client cert auth.

Reply:
Huh, this seems like a unintuitive interface decision - thank you very much for documenting it.  I don't feel particularly strongly about the check for the username - is it invalid to connect to a CouchDB without basic auth?  If so, then I think it makes sense to fail if there is no username.  Otherwise I am ok with this as is.

Reply:
The CouchDB container accepts the `COUCHDB_USER` and `COUCHDB_PASSWORD` parameters but they are not mandatory. In case they are omitted, the CouchDB can be accessed without any authentication.

https://hub.docker.com/_/couchdb/

Reply:
Ah ok thanks for explaining.  I'm ok with this snippet as is, thank you!